core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -0,0 +1,533 @@
+                },
+                SessionProtocol.HTTP, SessionProtocol.HTTPS,
+                SessionProtocol.H1, SessionProtocol.H1C,
+                SessionProtocol.H2, SessionProtocol.H2C);
+        connectTimeoutMillis = (Integer) baseBootstrap.config().options()
+                                                      .get(ChannelOption.CONNECT_TIMEOUT_MILLIS);
+    }
+
+    /**
+     * Returns an array whose index signifies {@link SessionProtocol#ordinal()}.
+     */
+    private static <T> T[] newEnumMap(Class<?> elementType,
+                                      Function<SessionProtocol, T> factory,
+                                      SessionProtocol... allowedProtocols) {
+        @SuppressWarnings("unchecked")
+        final T[] maps = (T[]) Array.newInstance(elementType, SessionProtocol.values().length);
+        // Attempting to access the array with an unallowed protocol will trigger NPE,
+        // which will help us find a bug.
+        for (SessionProtocol p : allowedProtocols) {
+            maps[p.ordinal()] = factory.apply(p);
+        }
+        return maps;
+    }
+
+    private Bootstrap getBootstrap(SessionProtocol desiredProtocol) {
+        return bootstraps[desiredProtocol.ordinal()];
+    }
+
+    @Nullable
+    private Deque<PooledChannel> getPool(SessionProtocol protocol, PoolKey key) {
+        return pool[protocol.ordinal()].get(key);
+    }
+
+    private Deque<PooledChannel> getOrCreatePool(SessionProtocol protocol, PoolKey key) {
+        return pool[protocol.ordinal()].computeIfAbsent(key, k -> new ArrayDeque<>());
+    }
+
+    @Nullable
+    private Future<Channel> getPendingConnection(SessionProtocol desiredProtocol, PoolKey key) {
+        return pendingConnections[desiredProtocol.ordinal()].get(key);
+    }
+
+    private void setPendingConnection(SessionProtocol desiredProtocol, PoolKey key, Future<Channel> future) {
+        pendingConnections[desiredProtocol.ordinal()].put(key, future);
+    }
+
+    private void removePendingConnection(SessionProtocol desiredProtocol, PoolKey key) {
+        pendingConnections[desiredProtocol.ordinal()].remove(key);
+    }
+
+    /**
+     * Acquires a {@link Channel} which is matched by the specified condition.
+     */
+    CompletableFuture<PooledChannel> acquire(SessionProtocol desiredProtocol, PoolKey key) {
+        final CompletableFuture<PooledChannel> promise = new CompletableFuture<>();
+        acquire(desiredProtocol, key, promise);
+        return promise;
+    }
+
+    private void acquire(SessionProtocol desiredProtocol, PoolKey key,
+                         CompletableFuture<PooledChannel> promise) {
+        final PooledChannel ch = acquireNow(desiredProtocol, key);
+        if (ch != null) {
+            promise.complete(ch);
+            return;
+        }
+
+        // Try to use the pending connection to avoid creating an extra connection.
+        final Future<Channel> pendingChannel = getPendingConnection(desiredProtocol, key);
```

Comment:
Should we check `desiredProtocol.isMultiplex()`?

Reply:
We do not know the protocol will be multiplex or not at this point if `desiredProtocol` is `HTTP` or `HTTPS`, but yeah we can do something after the pending connection is ready. Updated the logic and extracted it into `usePendingAcquisition()`. Please review.